pep-0692.rst

```diff
@@ -0,0 +1,344 @@
+        posonlyargs=[],
+        args=[],
+        kwonlyargs=[],
+        kw_defaults=[],
+        kwarg=arg(
+          arg='kwargs',
+          annotation=DoubleStarred(
+            value=Name(id='Movie', ctx=Load()),
+            ctx=Load())),
+        defaults=[]),
+      body=[
+        Expr(
+          value=Constant(value=Ellipsis))],
+      decorator_list=[])],
+    type_ignores=[])
+
+The runtime annotations should be consistent with the AST. Continuing the previous example:
+
+::
+
+    >>> def foo(**kwargs: **Movie): ...
+    ...
+    >>> foo.__annotations__
+    {'kwargs': **Movie}
+
+In order to make the runtime behaviour consistent with the AST as shown above we propose to:
+
+1. Add a new ``_UnpackedTypedDict`` class to the ``typing`` module and include it in ``_TypedDictMeta`` as a ``_unpacked`` property.
+2. Add a special ``__unpack__`` method to the ``_TpedDictMeta`` class that would return an instance of the ``_UnpackedTypedDict``.
+
+Then, specyfing ``def foo(**kwargs: **Movie): ...`` would call an ``__unpack__`` special method on ``Movie``.
+
+To be more precise the changes in the ``typing`` module would look as follows:
+
+::
+
+    class _UnpackedTypedDict:
+        def __init__(self, name):
+            self._name = name
+
+        def __repr__(self):
+            return '**' + self._name 
+    
+    class _TypedDictMeta(type):
+        def __init__(self, name):
+            self._name = name
+            self._unpacked = _UnpackedTypedDict(name)
+
+        def __unpack__(self):
+            return self._unpacked
+
+
+Implications
+------------
+
+With this change some of the allowed annotations would make no sense, for example:
+
+::
+
+    def foo(**kwargs: **int): ...
+
+Those kind of issues should be addressed by static type checkers. Specifically, the new double asterisk syntax should be only used with instances of ``TypedDict``. At runtime the objects that do not have an ``__unpack__`` method should behave as below:
+
+::
+
+    >>> def foo(**kwargs: **int): ...
+    ...
+    AttributeError: type object 'int' has no attribute '__unpack__'
+
+Backwards Compatibility
+-----------------------
+
+Using the double asterisk operator for annotating ``**kwargs`` is a syntax that would be only available in new versions of Python. PEP 646 dealt with the similar problem and its authors introduced a new type operator ``Unpack``. For the purposes of this PEP, the proposition is to reuse ``Unpack`` for more precise ``**kwargs`` typing. For example:
+
+::
+
+    def foo(**kwargs: Unpack[Movie]) -> None: ...
+
+There are several reasons for reusing PEP 646's ``Unpack``. Firstly, the name is quite suitable and intuitive for the ``**kwargs`` typing use case as the keywords arguments are "unpacked" from the ``TypedDict``. Secondly, there would be no need to introduce any new special forms. Lastly, the use of ``Unpack`` for the purposes described in this PEP does not interfere with the use cases described in PEP 646.
```

Comment:
I don't think there is a need for that - in the case of `Unpack` the AST should be consistent with the runtime annotations. The argument for doing that would be to keep `Unpack[Movie]` and `**kwargs: **Movie` consistent, but then `Unpack` would probably have to call that as well when being used in different contexts - it feels like it might be worth it to keep it as so that `Unpack` had the flexibility for other use cases?